### PR TITLE
Add corona to Gitlab CI, CI hardening, and documentation

### PR DESCRIPTION
**Pull request by adrienbernede**
This PR does the following:
- Add corona to Gitlab CI.
- Change the location of the dependencies installation.
  - Add the hash to the dependency path so that we avoid conflicts and can install more.
  - Use a symlink so that it's easier to switch between two dependency trees (think spack update).
- Make the mechanism to reproduce CI builds locally more robust, and document it.

Note: Part of the changes rely on the change of the `mfem-uberenv` commit we point too. You may want to review those changes too.

This PR is based upon the work done in https://github.com/mfem/mfem/pull/2425 to update the Spack package.

Outstanding TODOs:
- [x] Fix the compilation issue in Navier miniapp.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2428](https://github.com/mfem/mfem/pull/2428) | @adrienbernede | @tzanio | @camierjs + @YohannDudouit + @pazner | 08/17/21 | 09/13/21 | 09/14/21 | |

### CODE REVIEW

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
The next step is to add baseline tests for corona and lassen, with proper report in the CI.
I would like to do that in a subsequent PR, because it will introduce revamping of the CI, which would make the PR here harder to review.
Plus, this add a build and test on corona, which is already helpful.
Note that we had to fix the slurm issue for corona too, new slurm was rolled-out earlier on this system.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by tzanio** [comment]
Resolve conflicts in `tests/gitlab/get_mfem_uberenv`?

**Comment by adrienbernede** [comment]
LGTM

**Comment by tzanio** [comment]
@camierjs, @YohannDudouit and @pazner -- it will be good to merge this soon, can you please prioritize the review?

File: .gitlab/corona.yml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+# GitLab pipelines configurations for the Corona machine at LLNL
+
+.on_corona:
+  tags:
+    - shell
+    - corona
+  rules:
+    # Don’t run corona jobs if...
+    # Note: This makes corona an "opt-in" machine. Gitlab instances where
+    # corona jobs should run should set "ON_CORONA" to "ON".
```
Comment (pazner) [inline .gitlab/corona.yml:21]: Some more detailed instructions would be valuable here. I think I figured it out (adding the `ON_CORONA` variable in the GitLab repository settings), but it wasn't the most clear to me.

Maybe also note that this variable in set for our "LGTM" repo. (Incidentally, we really should change the trigger to something else, e.g. "LC GitLab: test")

**Comment by pazner** [comment]
I tried running this on my "personal" GitLab testing repo. I don't have an account on Corona, so no surprise that those jobs failed.

For some reason, the Lassen job failed too (`[ERROR: failure of spack install/dev-build]`).

https://lc.llnl.gov/gitlab/pazner1/mfem_testing/-/pipelines/91595

File: .gitlab/corona.yml (hunk 1 above)
Comment (adrienbernede) [reply .gitlab/corona.yml:21]: 1st point: understood. 👍🏻 
2nd point: I will remind gitlab admins about this. ;)

**Comment by adrienbernede** [comment]
> I tried running this on my "personal" GitLab testing repo. I don't have an account on Corona, so no surprise that those jobs failed.
> 
> For some reason, the Lassen job failed too (`[ERROR: failure of spack install/dev-build]`).
> 
> https://lc.llnl.gov/gitlab/pazner1/mfem_testing/-/pipelines/91595

Please hit "retry". The job should pass.
I reported this to Greg Becker, but it is not reproducible in my experience. Let me know if you ever run into that again.

**Comment by adrienbernede** [comment]
@camierjs, @YohannDudouit and @pazner 

Please see WIP documentation about how to reproduce those jobs locally here:
https://github.com/mfem/mfem/blob/aec5eb791f8affb5a22ed2ef45efc9fa8d02ff39/tests/gitlab/reproduce-ci-jobs-interactively.md

**Comment by pazner** [comment]
> > I tried running this on my "personal" GitLab testing repo. I don't have an account on Corona, so no surprise that those jobs failed.
> > For some reason, the Lassen job failed too (`[ERROR: failure of spack install/dev-build]`).
> > https://lc.llnl.gov/gitlab/pazner1/mfem_testing/-/pipelines/91595
> 
> Please hit "retry". The job should pass.
> I reported this to Greg Becker, but it is not reproducible in my experience. Let me know if you ever run into that again.

I tried re-running the job, and it failed again. Should I re-run the entire pipeline?

**Comment by adrienbernede** [comment]
> > > I tried running this on my "personal" GitLab testing repo. I don't have an account on Corona, so no surprise that those jobs failed.
> > > For some reason, the Lassen job failed too (`[ERROR: failure of spack install/dev-build]`).
> > > https://lc.llnl.gov/gitlab/pazner1/mfem_testing/-/pipelines/91595
> > 
> > 
> > Please hit "retry". The job should pass.
> > I reported this to Greg Becker, but it is not reproducible in my experience. Let me know if you ever run into that again.
> 
> I tried re-running the job, and it failed again. Should I re-run the entire pipeline?

This should be fixed now! (at least temporarily, I will submit a fix).

**Comment by pazner** [comment]
> @camierjs, @YohannDudouit and @pazner
> 
> Please see WIP documentation about how to reproduce those jobs locally here:
> https://github.com/mfem/mfem/blob/aec5eb791f8affb5a22ed2ef45efc9fa8d02ff39/tests/gitlab/reproduce-ci-jobs-interactively.md

Following those instructions on my MacBook, I get
```
==> Error: No compilers with spec gcc@6.1.0 found for operating system catalina and target skylake.
```

I tried with different `SPEC` configurations (`gcc@10.2.0`, `clang`) but no luck.

**Comment by adrienbernede** [comment]
> > @camierjs, @YohannDudouit and @pazner
> > Please see WIP documentation about how to reproduce those jobs locally here:
> > https://github.com/mfem/mfem/blob/aec5eb791f8affb5a22ed2ef45efc9fa8d02ff39/tests/gitlab/reproduce-ci-jobs-interactively.md
> 
> Following those instructions on my MacBook, I get
> 
> ```
> ==> Error: No compilers with spec gcc@6.1.0 found for operating system catalina and target skylake.
> ```
> 
> I tried with different `SPEC` configurations (`gcc@10.2.0`, `clang`) but no luck.

Sorry for the confusion, I said "locally" instead of "interactively", and by interactively, I mean on the same machines as in CI.

**Comment by adrienbernede** [comment]
Note that it is quite hard to create a generic configuration of Spack for any MacOS developer machine, unless we install _everything_ from scratch.

**Comment by pazner** [comment]
> Sorry for the confusion, I said "locally" instead of "interactively", and by interactively, I mean on the same machines as in CI.

Ok, got it. Yeah, I was indeed thrown off by "locally".

**Review by pazner — APPROVED** [review]
This looks good to me!

I don't have an account on Corona, so I would suggest that one of the reviewers who does should check that it all works with @adrienbernede's instructions for running the CI tests interactively, and possibly also in a personal GitLab repo.

**Comment by adrienbernede** [comment]
And please ask for any questions. The instructions are on another branch that comes next. So there are some slight changes that may results in the instructions not being exactly correct unless you use that branch.

I want you to become familiar with this. I truly think this can make you life much easier to build dependencies. And it will make it easier to share the exact same build instructions too (think interns)!

**Comment by pazner** [comment]
This has to do with the MFEM data repo and #2392, but I haven't had a chance to look closely.

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
@pazner @camierjs @YohannDudouit review can resume now ;)

**Comment by camierjs** [comment]
Looks good to me! Works great on `corona` now!

**Review by camierjs — APPROVED** [review]

File: tests/gitlab/build_and_test [hunk 2]
```diff
@@ -14,41 +14,125 @@
...
+  echo ""
+  echo " --build-only"
+  echo "    Only build MFEM, based on the existing configuration."
+  echo ""
+  echo " --test-only"
+  echo "    Only test MFEM, based on the existing configuration."
+  echo ""
+  echo " --data"
+  echo "    Make running tests with \"--data\" a requirement: will fail in the"
+  echo "    data directory is not present in mfem root directory."
```
Comment (pazner) [inline tests/gitlab/build_and_test:53]:
```suggestion
  echo "    data directory is not present in the parent of the mfem root directory."
```

**Comment by pazner** [comment]
> LGTM

Why don't I see the ci/gitlab check on the latest commit before your LGTM (`445c7de`)?

File: tests/gitlab/build_and_test (hunk 2 above)
Comment (adrienbernede) [reply tests/gitlab/build_and_test:53]: Good catch!

**Comment by adrienbernede** [comment]
@pazner that is a good point: There was a bug introduced by the latest update of LC Gitlab, and the mirroring is not working.
I noticed it last Friday (as well as @v-dobrev) but the bug was likely introduced on last Tuesday or Wednesday (effective update time), hence the absence of the report here, which I didn't even noticed.

You can see the pipeline working in my personal repo here: https://lc.llnl.gov/gitlab/bernede1/mfem/-/pipelines/94601

The bug Gitlab bug should be fixed today (https://mattermost.energy.gov/asc-trilab/pl/mef7mq6p1pd9uf4qpo33ewd8jo).

**Review by pazner — APPROVED** [review]
Looks great, nice contribution @adrienbernede!

Maybe let's wait until the LC GitLab mirroring bug is fixed to check the LGTM one last time :)

**Comment by adrienbernede** [comment]
LGTM

**Comment by pazner** [comment]
> LGTM

It still doesn't show up for me 🤔

**Comment by adrienbernede** [comment]
@pazner I know... I have my favorite Gitlab admin working on it :D

**Comment by adrienbernede** [comment]
LGTM

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by adrienbernede** [comment]
LGTM

**Pull request merged**